### DISCUSSION TITLE
How to convert/set bit depth from 1 to 8 bit

### DISCUSSION THREAD

**Discussion by ademmler** [libvips/libvips#2777] [Q&A] [answered]
How can I set / convert 1bit BW images to 8Bit Greyscale using libvips and c++.

**Comment by jcupitt** [comment, marked as answer]
Just set `bitdepth`, eg.:

```
$ vips copy mono.png x.tif[bitdepth=1]
$ tiffinfo x.tif
TIFF Directory at offset 0x372744 (5b008)
  Image Width: 1450 Image Length: 2048
  Resolution: 28.35, 28.35 pixels/cm
  Bits/Sample: 1
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
$ 
```

Same for PNG and GIF.

**Reply by jcupitt** [reply]
libvips defaults to 8 bit, so you don't need to do anything going the other way:

```
$ vips copy x.tif x.png
$ vipsheader x.png
x.png: 1450x2048 uchar, 1 band, b-w, pngload
```

**Reply by ademmler** [reply]
Hi jcupitt, 
I see and this is what I got from documentation also.

I tried to use 1bit tiffs within my "spot colors example" #2037 but I get
repettetive errors (below samples) while writing the result and "abortrap 6" reached finally.

(process:5101): VIPS-WARNING **: 11:18:51.671: Unknown field with tag 32781 (0x800d) encountered
(process:5101): VIPS-WARNING **: 11:18:51.681: error in tile 0 x 127
libc++abi.dylib: terminating with uncaught exception of type vips::VError: tiff2vips: out of order read -- at line 2835, but line 0 requested
tiff2vips: out of order read -- at line 2835, but line 127 requested

I guess I need some conversion or hint for vips to merge.

**Reply by jcupitt** [reply]
Oh, strange, it should work. You'd need to post an example I could check.

**Reply by ademmler** [reply]
I have moved the test files below ...

**Reply by ademmler** [reply]
Thx for checking this out.
Here is a sample for download.
I created lowres 1bit files - to keep the upload it small.
It will look a bit ugly - but for testing this is ok
[test1bit.zip](https://github.com/libvips/libvips/files/8582083/test1bit.zip)

**Reply by ademmler** [reply]
Hello @jcupitt,

Anny findings on this issue?

Regards Alexander

**Reply by jcupitt** [reply]
Sorry, I've been overwhelmed with work this week.

I read your code carefully, and you just need to remove this line:

```C++
        cmyk_layers.invert().write_to_file("cmyk.tif");          
```

If you open an image in sequential mode, you can only read it once. Your program read the CMYK images once to generate `cmyk.tif`, then did the composite, then tried to read again to generate `merged.tif`. Since the inputs had already been read, you had the error.

If you want to read again, you need to open again. So another option would be this:

```C++
// compile with:
//
//   g++ -std=c++11 -g -Wall -o recombine recombine.cpp `pkg-config vips-cpp --cflags --libs`

#include <stdlib.h>
#include <vips/vips8>

using namespace vips;

VImage
vopen(const char *filename)
{   
    const VipsAccess access = VIPS_ACCESS_SEQUENTIAL;
    // const VipsAccess access = VIPS_ACCESS_RANDOM;

    return VImage::new_from_file(filename, VImage::option()
            ->set ("access", access));
}

int
main( int argc, char **argv )
{
    if(VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    //Load images
    VImage c = vopen("testpage-Cyan.tif");
    VImage m = vopen("testpage-Magenta.tif");
    VImage y = vopen("testpage-Yellow.tif");
    VImage k = vopen("testpage-Black.tif");
    VImage spot1 = vopen("testpage-Spot1.tif");
    VImage spot2 = vopen("testpage-Spot2.tif");

    //Create CMYK image
    VImage cmyk_layers = VImage::bandjoin({c, m, y, k})
        .copy(VImage::option()
            ->set("interpretation", VIPS_INTERPRETATION_CMYK));

    cmyk_layers.invert().write_to_file("cmyk.tif");


    c = vopen("testpage-Cyan.tif");
    m = vopen("testpage-Magenta.tif");
    y = vopen("testpage-Yellow.tif");
    k = vopen("testpage-Black.tif");
    cmyk_layers = VImage::bandjoin({c, m, y, k}) 
        .copy(VImage::option()
            ->set("interpretation", VIPS_INTERPRETATION_CMYK));

    VImage sp1_layers = cmyk_layers
        .new_from_image({0, 255, 255, 0}) 
        .bandjoin(spot1)
        .invert();

    VImage sp2_layers = cmyk_layers
        .new_from_image({255, 0, 255, 0}) 
        .bandjoin(spot2)
        .invert();

    //Merge all together
    VImage merged = VImage::composite({cmyk_layers, sp1_layers, sp2_layers},
        {VIPS_BLEND_MODE_MULTIPLY, VIPS_BLEND_MODE_MULTIPLY}, VImage::option()
            ->set ("compositing-space", "cmyk"))
        .invert()
        .extract_band(0, VImage::option()
            ->set ("n", 4));

    merged.write_to_file("merged.tif");

    return 0;
}
```

Which also works.

**Reply by jcupitt** [reply]
(but I suggest you just remove the write of `cmyk.tif`, I guess this was debugging code)

**Reply by ademmler** [reply]
You are right it was for debugging - hence I have learned a lot about libvips now - thank you so much! I really appreciate your support.

**Reply by ademmler** [reply]
I tried to convert the CMYK image first into RGB using an ICC Profile and than add the spotcolor layers as RGB layers. The reason is that I would like to define the spmotcolor in RGB (or later even as LAB values)

Here is what I changed in main() to achieve this. The result is indeed a RGB image tagged with my profile.
But I have noticed that the color information I set for sp1_layers & sp2_layers is still interpreted as CMYK.

I tried to 
a) use ".colorspace()" for rgb_image and the sp_layers.
b) add ".extract_band(0, VImage::option()->set ("n", 3))" the sp_layers,
but both did not worked.

What do I need to to enable RGB as color definition.

Regards Alex

```
//Load images
    VImage c = vopen("testpage-Cyan.tif");
    VImage m = vopen("testpage-Magenta.tif");
    VImage y = vopen("testpage-Yellow.tif");
    VImage k = vopen("testpage-Black.tif");
    VImage spot1 = vopen("testpage-Spot1.tif");
    VImage spot2 = vopen("testpage-Spot2.tif");

    //Create CMYK image
    VImage rgb_image = VImage::bandjoin({c, m, y, k})
        .copy(VImage::option()->set("interpretation", VIPS_INTERPRETATION_CMYK))
        .invert()
        .icc_transform("LStar-RGB.icc");
        
    VImage sp1_layers = rgb_image
        .new_from_image({255, 0, 0})
        .bandjoin(spot1)
        .invert();

    VImage sp2_layers = rgb_image
        .new_from_image({0, 0, 255})
        .bandjoin(spot2)
        .invert();

    //Merge all together
    VImage merged = VImage::composite({rgb_image, sp1_layers, sp2_layers},
        {VIPS_BLEND_MODE_MULTIPLY, VIPS_BLEND_MODE_MULTIPLY})
        .extract_band(0, VImage::option()
            ->set ("n", 3));

    merged.write_to_file("merged.tiff");
```

**Reply by ademmler** [reply]
@jcupitt  Hi John, I guess you are very busy - I do not want o bother - just hoping you find time to tell me how to "use RGB" and LAB" instead of CMYK for the spot colors (CMYK colorspace is to small for many colors). I tried a lot but the figures where always interpreted as CMY values.  Regards Alex

**Reply by jcupitt** [reply]
Hi @ademmler, sorry, I got distracted.

You are writing a CMYK image, so your spot colours will need to be in the CMYK gamut, won't they?

**Reply by ademmler** [reply]
Ok - intially my question was targeting for this - but my final goal is this:
First combine CMYK channels -> transfer to RGB or LAB (see code above) -> add spot colors as RGB or LAB values -> save LAB or RGB image.  Therefore I do a .icc_transform("LStar-RGB.icc"); on the CMYK image - than I try to add the spotoclors  - the result is an RGB image. But the colors I passed to sp1. and sp2. are interpreted as CMYK not as RGB values. What I am missing here?

**Reply by jcupitt** [reply]
Ah, OK, this seems to work for me:

```C++
// compile with:
//
//   g++ -std=c++11 -g -Wall -o recombine-rgb recombine_rgb.cpp `pkg-config vips-cpp --cflags --libs`

#include <stdlib.h>
#include <vips/vips8>

using namespace vips;

VImage
vopen(const char *filename)
{
    const VipsAccess access = VIPS_ACCESS_SEQUENTIAL;

    return VImage::new_from_file(filename, VImage::option()
            ->set ("access", access));
}

int
main( int argc, char **argv )
{
    if(VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    VImage c = vopen("testpage-Cyan.tif");
    VImage m = vopen("testpage-Magenta.tif");
    VImage y = vopen("testpage-Yellow.tif");
    VImage k = vopen("testpage-Black.tif");
    VImage spot1 = vopen("testpage-Spot1.tif");
    VImage spot2 = vopen("testpage-Spot2.tif");

    // Create CMYK image
    // these planes have white for no-ink, so we need to invert (cmyk uses 255
    // to mean lots of ink)
    VImage cmyk_layers = VImage::bandjoin({c, m, y, k})
        .copy(VImage::option()
            ->set("interpretation", VIPS_INTERPRETATION_CMYK))
        .invert();

    // to RGB
    VImage rgb_layers = cmyk_layers.colourspace(VIPS_INTERPRETATION_sRGB);

    // we have to invert the spot mask, since it uses 0 to mean lots of ink,
    // but rgb uses 0 to mean black
    VImage sp1_layers = rgb_layers
        .new_from_image({255, 255, 0})
        .bandjoin(spot1.invert());

    VImage sp2_layers = rgb_layers
        .new_from_image({255, 0, 255})
        .bandjoin(spot2.invert());

    //Merge all together
    VImage merged = VImage::composite({rgb_layers, sp1_layers, sp2_layers},
        {VIPS_BLEND_MODE_MULTIPLY, VIPS_BLEND_MODE_MULTIPLY});

    merged.write_to_file("merged.tif");

    return 0;
}
```

I don't know if that will overlays inks in the way you expect. You might want to try OVER instead of MULTIPLY.

**Reply by ademmler** [reply]
Hi @jcupitt  - thx for this hints. Yes basically this works. but I have more questions:

1. sRGB is a small RGB colorspace - I would like to use either some like L-Star or Adobe RGB. 
That was my aim using .icc_transform("LStar-RGB.icc"); instead of cmyk_layers.colourspace(VIPS_INTERPRETATION_sRGB);
Is there a reason why you have changed it?

2. I tried to do the same with LAB as a final colorspace. But the result is still RGB. Again I am doing wrong.
Somehow I have difficulties getting the concept of libvips into my brain.
```
// to LAB
    VImage rgb_layers = cmyk_layers.colourspace(VIPS_INTERPRETATION_LAB);

    // we have to invert the spot mask, since it uses 0 to mean lots of ink,
    // but rgb uses 0 to mean black
    VImage sp1_layers = rgb_layers
        .new_from_image({54, 80, 70})
        .bandjoin(spot1.invert());

    VImage sp2_layers = rgb_layers
        .new_from_image({97, -15, 92})
        .bandjoin(spot2.invert());

    //Merge all together
    VImage merged = VImage::composite({rgb_layers, sp1_layers, sp2_layers},
        {VIPS_BLEND_MODE_MULTIPLY, VIPS_BLEND_MODE_MULTIPLY});

    merged.write_to_file("merged.tif");
```

**Reply by jcupitt** [reply]
You can use float srgb, so values >255 and <0. It should let you represent any colour. Just cast to float.

```C++
    // to RGB
    VImage rgb_layers = cmyk_layers
        .colourspace(VIPS_INTERPRETATION_sRGB)
        .cast(VIPS_FORMAT_FLOAT);
```

`composite` doesn't work well for LAB, unfortunately. `ab` are +/1 128, but how does that work with 0 - 255 alpha? It's not well defined.

**Reply by jcupitt** [reply]
... or use your l-star profile, of course.

**Reply by ademmler** [reply]
Ok I see - because the final step is to go to the display RGB profile (for best viewing result) I will try to use L-Star in-between (as a calibrated colorspace).  I guess I can do the final conversion also in vips.
Have I understood correctly - that you use LCMS inside vips for all ICC stuff?

**Reply by jcupitt** [reply]
It's a mix, but lcms is used for conversions using profiles.

**Reply by ademmler** [reply]
Seems like conversion from sRGB to other RGB is not possible - I get this error:
(process:3241): VIPS-WARNING **: 15:10:02.420: profile incompatible with image

```
merged = VImage::composite({rgb_layers, sp1_layers, sp2_layers},
        {VIPS_BLEND_MODE_MULTIPLY, VIPS_BLEND_MODE_MULTIPLY});			

VImage final_image = merged.icc_transform("LStar-RGB.ICC");
```

**Reply by jcupitt** [reply]
It should work. Convert the CMYK to lstar, eg.:

```C++
// compile with:
//
//   g++ -std=c++11 -g -Wall -o recombine-rgb recombine_rgb.cpp `pkg-config vips-cpp --cflags --libs`

#include <stdlib.h>
#include <vips/vips8>

using namespace vips;

VImage
vopen(const char *filename)
{
    const VipsAccess access = VIPS_ACCESS_SEQUENTIAL;

    return VImage::new_from_file(filename, VImage::option()
            ->set ("access", access));
}

int
main( int argc, char **argv )
{
    if(VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    VImage c = vopen("testpage-Cyan.tif");
    VImage m = vopen("testpage-Magenta.tif");
    VImage y = vopen("testpage-Yellow.tif");
    VImage k = vopen("testpage-Black.tif");
    VImage spot1 = vopen("testpage-Spot1.tif");
    VImage spot2 = vopen("testpage-Spot2.tif");

    // Create CMYK image
    // these planes have white for no-ink, so we need to invert (cmyk uses 255
    // to mean lots of ink)
    VImage cmyk_layers = VImage::bandjoin({c, m, y, k})
        .copy(VImage::option()
            ->set("interpretation", VIPS_INTERPRETATION_CMYK))
        .invert();

    // to RGB
    VImage rgb_layers = cmyk_layers
        .icc_transform("LStar-RGB-v4.icc");

    // we have to invert the spot mask, since it uses 0 to mean lots of ink,
    // but rgb uses 0 to mean black
    VImage sp1_layers = rgb_layers
        .new_from_image({255, 255, 0})
        .bandjoin(spot1.invert());

    VImage sp2_layers = rgb_layers
        .new_from_image({255, 0, 255})
        .bandjoin(spot2.invert());

    //Merge all together
    VImage merged = VImage::composite({rgb_layers, sp1_layers, sp2_layers},
        {VIPS_BLEND_MODE_MULTIPLY, VIPS_BLEND_MODE_MULTIPLY});

    merged.write_to_file("merged.tif");

    return 0;
}
```

To generate:

![merged](https://user-images.githubusercontent.com/580843/168294902-90f77c83-c5e6-4669-a1f9-af1715bcdb1d.jpg)

The `{255, 255, 0}` are in lstar space.

**Reply by ademmler** [reply]
Hi - thank you for helping me with that again!
**The honor is again for you - perfect support!**

?How can I mark just your last part to be the full answer?

**Comment by ademmler** [comment]
Hi, I am back. I am trying to run my sample code with very large tiffs.
I reduce the example to CMYK colors only.

... shortened
 VImage cmyk_layers = VImage::bandjoin({c, m, y, k})
        .copy(VImage::option()
            ->set("interpretation", VIPS_INTERPRETATION_CMYK))
        .invert();

cmyk_layers.write_to_file("möget.tif");
... shortend

I have 4 tiffs - one for each color - like this one.

```
TIFF Directory at offset 0x8 (8)
  Subfile Type: (0 = 0x0)
  Image Width: 116977 Image Length: 271560
  Resolution: 2400, 2400 pixels/inch
  Bits/Sample: 1
  Compression Scheme: LZW
  Photometric Interpretation: min-is-white
  Samples/Pixel: 1
  Rows/Strip: 229
  Planar Configuration: single image plane
  ImageDescription: Black
  Software: LnTIFF
  Tag 32781: 1. 631d1d9b-b1be-4a17-be9e-9fcf2bebb098.jdf (K)

```
The resulting tiff has this characteristics:

```
TIFF Directory at offset 0x1d95ab8130 (127065096496)
  Image Width: 116977 Image Length: 271560
  Resolution: 2400, 2400 pixels/inch
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: separated
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 4
  Rows/Strip: 128
  Planar Configuration: single image plane
  ImageDescription: Cyan
  InkSet: 1

```

1. The resulting image can't be opened or displayed by an image viewer like Photoshop.
What should I change to get a more compatible tiff.

2. The size is immense (128 GB). Can I reduce the images on the fly or before merging?

3. How to speed up the whole process?

Regards Alexander

**Reply by jcupitt** [reply]
You can set LZW compression on the output image, perhaps:

```C++
    merged.write_to_file("merged.tif", VImage::option()
            ->set("compression", VIPS_FOREIGN_TIFF_COMPRESSION_LZW));
```

or equivalently:

```C++
    merged.write_to_file("merged.tif[compression=lzw]");
```

**Reply by jcupitt** [reply]
... you might need to set the `bigtiff` option as well.

**Comment by ademmler** [comment]
Hi, thx for your response and hint. I tried your suggestion together with downsampling.
But it is still extremely slow.  What would be the best way - order- to downsample and convert to Standard 8bit CMYK?
```

#include <stdlib.h>
#include <vips/vips8>

using namespace vips;

int
main( int argc, char **argv )
{
    if(VIPS_INIT(argv[0]))
        vips_error_exit(NULL);
    
    //Load images
    VImage c = VImage::new_from_file("testpage(Cyan).tif",
                                     VImage::option()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    VImage m = VImage::new_from_file("testpage(Magenta).tif",
                                     VImage::option()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    VImage y = VImage::new_from_file("testpage(Yellow).tif",
                                     VImage::option()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    VImage k = VImage::new_from_file("testpage(Black).tif",
                                     VImage::option()->set ("access", VIPS_ACCESS_SEQUENTIAL));
                                     
	//Reduce images
	c.resize(0.25);
	m.resize(0.25);
	y.resize(0.25);	
	k.resize(0.25);                                     
          
    //Create CMYK image
    VImage image = VImage::bandjoin({c, m, y, k})
            .copy(VImage::option()
                    ->set("interpretation", VIPS_INTERPRETATION_CMYK))
                    .invert();
        
	image.write_to_file("merged.tif", VImage::option()
            ->set("compression", VIPS_FOREIGN_TIFF_COMPRESSION_LZW)
            ->set("bigtiff", true)
            ->set("tile", true)
            ->set("pyramid", true)
            );
    
    return 0;
}

```

**Reply by jcupitt** [reply]
> ->set("compression", VIPS_FOREIGN_TIFF_COMPRESSION_LZW)

Unfortunately LZW compression in libtiff is both slow and single-threaded. You could try saving without compression -- the files will be uselessly large, but it might confirm where the bottleneck is.

You could also bandjoin *before* the resize, it should help a little.

**Reply by jcupitt** [reply]
And obviously you need:

```C++
c = c.resize(0.25)
```

I expect that was just a typo.

**Reply by jcupitt** [reply]
... so I would write:

```C++
    VImage c = VImage::new_from_file("testpage(Cyan).tif",
                                     VImage::option()->set ("access", "sequential"));
    VImage m = VImage::new_from_file("testpage(Magenta).tif",
                                     VImage::option()->set ("access", "sequential"));
    VImage y = VImage::new_from_file("testpage(Yellow).tif",
                                     VImage::option()->set ("access", "sequential"));
    VImage k = VImage::new_from_file("testpage(Black).tif",
                                     VImage::option()->set ("access", "sequential"));
          
    VImage image = VImage::bandjoin({c, m, y, k})
            .copy(VImage::option()->set("interpretation", "cmyk"))
            .resize(0.25)
            .invert();
```
